packages/camera/src/components/Controls/hooks.js

```diff
@@ -1,56 +1,67 @@
-import { useCallback } from 'react';
+import { useCallback, useEffect, useState } from 'react';
 import Actions from '../../actions';
 
-const useHandlers = ({ onStartUploadPicture, onFinishUploadPicture, enableComplianceCheck }) => {
-  const capture = useCallback(async (state, api, event) => {
+const useHandlers = ({
+  onStartUploadPicture, onFinishUploadPicture, checkComplianceAsync,
+  enableComplianceCheck, state,
+}) => {
+  const [complianceToCheck, setComplianceToCheck] = useState([]);
+  /**
+   * Note(Ilyass): We removed the recursive function solution, because it takes too much time,
+   * instead we re-run the compliance one more time after 1sec of getting the first response
+   * */
+  const verifyComplianceStatus = (pictureId, compliances, currentId) => {
+    const hasTodo = Object.values(compliances).some((c) => c.status === 'TODO' || c.is_compliant === null);
+
+    if (hasTodo) {
+      setTimeout(async () => {
+        await checkComplianceAsync(pictureId, currentId);
+      }, 500);
+    }
+  };
+
+  useEffect(() => {
+    const index = complianceToCheck[0];
+
+    if (index && state.uploads.state[index].status === 'fulfilled') {
+      const pictureId = state.uploads.state[index].pictureId;
+      (async () => {
+        if (enableComplianceCheck) {
+          const result = await checkComplianceAsync(pictureId);
+          verifyComplianceStatus(pictureId, result.data.compliances, index);
+          setComplianceToCheck((prev) => prev.slice(1));
+          onFinishUploadPicture();
+        }
+      })();
+    }
+  }, [complianceToCheck, state.uploads.state]);
+
+  const capture = useCallback(async (customState, api, event) => {
+    const usedState = customState || state;
```

Comment:
Can you use the words `controlledState` and `unControlledState` like this `const state = controlledState || unControlledState;` This is what I understood. Maybe comments are necessary here.